---
title: "Data Description"
author: "Yan-Cheng Lin"
date: '2023-10-31'
bibliography: Reference.bib
output:
  html_document: default
  pdf_document: default
---
# Plant Material

We used four different maize populations developed from the two European flint maize landraces, Kemater Landmais Gelb (KE) and Petkuser Ferdinand Rot (PE) (@holker2019european, @mayer2020discovery). From each landrace, a doubled haploid (DH) population was derived directly from S0 plants, and a gamete capture (GC) population of S1:2 plants was generated by crossing S0 plants with the capture line, FV2, and subsequent selfing (@holker2022theoretical). The entire dataset comprises 1,417 landrace derived lines (DH_KE = 471, DH_PE = 402, GC_KE = 274, GC_PE = 270).

## Phenotypic Data and Heritability

Field design and trait assessment were described in detail by @holker2019european and @holker2022theoretical. In brief, the DH lines and GC-S1:2 lines were evaluated at two different locations, Roggenstein (ROG) and Einbeck (EIN), in Germany in two years (2017 and 2018). Line per se performance was assessed for five traits, early vigor at V6 stage (EV_V6, 1-9 score), plant height at V6 stage (PH_V6, cm), final plant height (PH_final, cm), female flowering time (DtSILK, days from sowing to 50% of plants in a plot silked) and severity of root lodging at R6 stage (RL_R6, 1-9 score). For all traits, we calculated adjusted means across the four environments as described in @holker2019european.

```{r}
Pheno <- read.table("../Data/PhenotypicData.txt",
                    sep = "\t", header = T)

head(Pheno)
table(substr(Pheno$Genotype, start = 1, stop = 5))

```

```{r}
H2 <- read.table("../Data/Heritability.txt",
                 sep = "\t", row.names = 1)
print(H2)

```

## Genotypic Data

All DH lines and GC-S1 plants were genotyped with the 600k Affymetrix Axiom Maize Array (@unterseer2014powerful). Quality filtering and imputation followed @holker2019european for DH populations and @mayer2020discovery for GC populations. Original dataset with vcf format is available on https://figshare.com/articles/dataset/Data_from_HoelkerMayer_et_al/17014421
The analysis of this study will start with the transformed genotypic dataset "GenotypicData.RData", which encode the genotype with gamete form in a matrix. Please contact corresponding author for dataset "GenotypicData.RData", if needed.

```{r}
load("../Data/GenotypicData.RData")

# Physical position of the 600k SNP array markers on reference genome B73_v4
head(map)

# Genotypic matrix, with markers in column and genotype in row.
# Individual sample were split into two gamete to maintain the chromosome phasing information.
# Genotype carrying the same SNP allele as FV2 (capture line), will be encode as 1, otherwise 0. 
geno.sel.gamete[1:5,1:5]

table(substr(rownames(geno.sel.gamete), start = 1, stop = 5))
```
